### DISCUSSION TITLE
New row violates row-level security policy for table "objects"

### DISCUSSION THREAD

**Discussion by esDotDev** [supabase/supabase#13344] [Questions]
# Bug report
- [x ] I confirm this is a bug with Supabase, not with my own application. 
- [ x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

My authenticated clients can not see existing buckets, or insert new files.

I have
* added read support to all storage.buckets for all authorized users
* added insert support to storage.objects for all authorized users
* given ALL access to users in their own top-level directory within the `users` bucket

Using the anon key I get RLS errors when trying to insert, and no buckets are listed. If I switch to the service key, everything works as intended.
```
Unhandled Exception: StorageException(message: new row violates row-level security policy for table "objects", statusCode: 401, error: Invalid JWT)
```
I have followed everything in the docs as far as I can see, and am not sure where to go from here.

## Screenshots

![image](https://user-images.githubusercontent.com/736973/229146681-da50f270-233b-4966-9f36-286f9dda320b.png)
![image](https://user-images.githubusercontent.com/736973/229146816-76acdb25-6e15-4faa-9c89-e0c266502052.png)

## Additional context

Using the flutter sdk.

**Comment by GaryAustin1** [comment]
This is unlikely a bug at this point and probably a set up or issue passing in a user jwt when you make the storage call.
Going to move this to discussions for now so maybe you can get some help.

You should show your upload call and insure you have a user jwt as the foldername in your path as your RLS is checking that.
You could also try TO anon and true for the policy.   If that works then you don't have a user jwt at the time of the upload.

**Reply by GaryAustin1** [reply]
I also don't see that you have asked about this in Github or Discord before posting it as a bug,  but certainly could have missed it.

**Reply by esDotDev** [reply]
Thanks Gary, I did search all the discussions and issues I could find. I did not try with discord as I never tend to get an answer there.

The upload call is
```
database.storage.from('users').upload('68d0cd7a-2aae-4f99-a980-d5f51019d9c8/image.png', File(...))
```
No matter what I seem to set my policies to, I can't get it to work. 

I agree it's unlikely to be a bug, but I'm not sure how I can further trouble-shoot this. There are no error messages returned other than `401 Bad JWT`

`_You could also try TO anon and true for the policy. If that works then you don't have a user jwt at the time of the upload._`

I'm not sure what this means.  I have an authenticated user, I just want to allow them to push some files into a bucket that is private to them. When I use the service key, everything works fine and the upload is stored in the correct folder 
![image](https://user-images.githubusercontent.com/736973/229162637-5f83d72c-264b-4845-9732-469f1eb73122.png)

I'm happy to implement this in the simplest way possible and move on, but I need some guidance here. I tried using the template "_Give users access to only their own a top level folder named as uid_", as that is exactly the behavior I want, but for some reason it's not working.

**Reply by GaryAustin1** [reply]
I am suggesting you temporarily change TO (edit I see it is called 'Target Roles' in the UI) to 'anon' instead of authenticated.  Then in the actual policy just set to true.  AS A TEST.  If that works with your current call then it means you don't have a user jwt when you make the call.  This is a very common case of it not working so trying to rule that out first.

**Reply by esDotDev** [reply]
Ah understand, it's called "target roles" in the UI, I didn't realize that corresponds to a `TO` statement in the code. 

I tried that and it didn't seem to help.

Also, I my user is logged in and authenticated, it should have a valid JWT?
![image](https://user-images.githubusercontent.com/736973/229164829-6159bd44-bf27-4321-9587-73ebe5da7bd3.png)

**Reply by GaryAustin1** [reply]
Also you say using the anon key it does not work with your current RLS.  Yet in the policy you are using auth.uid() to compare the users jwt uuid to your folder.  Are you not signing in a user with auth before making the call?   If not then you can't use auth.uid() in the policy or authenticated.

**Reply by GaryAustin1** [reply]
I notice in this code you also have upsert is true.  This means you will have to also have and meet select policy and potentially update policy.

**Reply by esDotDev** [reply]
I am signing in before making the upload calls. You can see in the screenshot above, I print the `auth.currentSession.accessToken` and it's valid.

Additionally, when I use the serviceKey, the image is uploaded to the correct folder, which is `/users/[userId]/imageName`. 

So I know the path is correct, I know I'm signed in. Just not sure what's wrong...

**Reply by GaryAustin1** [reply]
In the only policy you show you have just a WITH CHECK that only works for INSERT.   Not sure if that is for all your policies or just INSERT.  You will need USING policy for both SELECT and UPDATE.

**Reply by esDotDev** [reply]
I have enabled all the permissions, SELECT, INSERT, UPDATE and DELETE, using your templates, like this: 
![image](https://user-images.githubusercontent.com/736973/229166562-dd7ce603-c684-47a0-8b42-8fa40ab934e5.png)

Do I need to do more than that?

Also, as you can see, it is no longer letting me add new policies, listing a generic "Internal Server Error" :/

**Reply by GaryAustin1** [reply]
These conversations are much easier in Discord.  I'm sorry you have had bad luck there as most people seem to get some help.

I suggest you clear all your policies and start from scratch with NO roles selected and TRUE for all policies.  Then just set roles to 'authenticated' and work your way back.

**Reply by GaryAustin1** [reply]
By "Your templates" I assume you mean Supabase's templates (I am not a team member).  All I have suggested to you for templates is to use TRUE to work things starting simply.
Not sure on your Server Error.  Maybe the Dashboard has logged out.

**Reply by esDotDev** [reply]
You're correct the dashboard had logged me out.

Ok, so I took your advice and made a simple policy like:
```
CREATE POLICY "basic 1ufimg_0" ON storage.objects FOR SELECT TO public USING (true);
```
And it now works, but as soon as I add `Authenticated` as the target role, it stops working. Man I am confused...

**Reply by GaryAustin1** [reply]
If you change public to anon does it also work?
If so then for what ever reason your call from flutter does not have a user jwt being passed out, but just the original anon key.

**Reply by esDotDev** [reply]
Ya, it does work when I change to `anon`.

But it sure looks like I am fully authenticated: 
![image](https://user-images.githubusercontent.com/736973/229175568-b06a4fc9-7991-4adf-aebd-7da5fc81a402.png)

**Reply by GaryAustin1** [reply]
decode your access token with jwt.io
make sure it is a user token and has role authenticated and not expired.

Other than that something about flutter sending it out I guess.  The server is getting your anon jwt and decoding it, so don't think it is a server side issue.

And people are using flutter client with storage so would have to be a new issue if that is the cause.

**Reply by esDotDev** [reply]
Thanks for all your replies, I think things are clearing up a bit. It does look fine ya?
![image](https://user-images.githubusercontent.com/736973/229177025-ccc024c8-f9a4-44f6-bc98-196c85236158.png)

**Reply by GaryAustin1** [reply]
Seems OK to me. 7 days from now expire on a jwt is a bit long, but not expired.
I don't know anything about flutter and if you have a way to actually look at the authenticated header like we can in javascript on a browser.
That is about all that is left for not being correct and the client should handle that.

**Reply by GaryAustin1** [reply]
https://github.com/supabase/supabase-flutter/issues/376
In the ballpark...

**Reply by esDotDev** [reply]
Hm I don't think that's it as I don't store off the client anywhere, always using accessorts/methods to fetch the latest.

When I inspect traffic in Flutter, I do see:
![image](https://user-images.githubusercontent.com/736973/229182242-47370ac6-52ea-4746-b4d6-b93bf3a3feb2.png)

Which appears to be an anon-key
![image](https://user-images.githubusercontent.com/736973/229182371-4726fc4b-3137-4294-b2b3-9dfd251b3118.png)

**Reply by esDotDev** [reply]
Interestingly there are two network requests, the one to storage uses the anon-key, but the next one to database/rest API is using the authorized token: 
![image](https://user-images.githubusercontent.com/736973/229182756-b7b21127-a20b-462b-a805-5d17cf183651.png)

![image](https://user-images.githubusercontent.com/736973/229182837-df9c2272-5bc4-4e5c-9917-e34d6064dee3.png)

**Reply by GaryAustin1** [reply]
Yeah that agrees with your results.
Don't know enough about flutter / flutter client to help further.
Some how your storage client is not getting the user from your general client.
Make sure you are on latest code, or maybe back up a version if you are to see if it is a new bug.
I would expect to see reports of this if it was a general code issue, but don't see any in the forums in a quick search.

I suggest you open a new discussion on Discord or as an issue in https://github.com/supabase/supabase-flutter/issues specifically talking about the storage authenticated header being incorrect.   As the RLS storage stuff is just confusing the matter at this point.

**Reply by esDotDev** [reply]
Ok will do, thx again for the help, very appreciated!

**Reply by esDotDev** [reply]
It does seem like this fixes the issue, but no mention I can find in the docs about settings this:
```
database.storage.setAuth(database.auth.currentSession!.accessToken);
```

It looks like this is a bug in the Flutter SDK, when you resume a session, the `_handleTokenChanged` method skips or misses the call to these various API:
```
storage.setAuth(token!);
functions.setAuth(token);
realtime.setAuth(token);
```

**Reply by esDotDev** [reply]
I've logged a bug here, I'm fairly certain it's an issue in the SDK. Thanks again, I'm not stupid/crazy, YAY!
https://github.com/supabase/supabase-flutter/issues/424

**Reply by esDotDev** [reply]
Was fixed in this version of gotrue for dart, so just had to update to the latest version, now sending the correct JWT token again.
https://github.com/supabase/gotrue-dart/pull/133

**Comment by esDotDev** [comment]
The issue was the JWT token not getting set correctly on the `storage` client, was fixed in v1.5.7:
https://github.com/supabase/gotrue-dart/pull/133

**Reply by d3roch4** [reply]
I using `supabase_flutter: ^1.10.22` and the problem continue :(
